numpy-stubs/__init__.pyi

```diff
@@ -82,10 +87,39 @@ _DtypeLike = Union[
     Tuple[_DtypeLikeNested, _DtypeLikeNested],
 ]
 
+# Can be used as such `Union[Sequence[T], _ArrayLikeSeq]` as ndarray
+# is not a proper Sequence
+# TODO: Specify the to-be returned array type once ndarray is a Generic
+class _ArrayLikeSeq(Protocol):
+    shape: Tuple[int, ...]
+    def __array__(self, dtype: _DtypeLike = ...) -> ndarray: ...
+
 _NdArraySubClass = TypeVar("_NdArraySubClass", bound=ndarray)
 
 _ArrayLike = TypeVar("_ArrayLike")
 
+_Generic = TypeVar("_Generic", bound=generic)
+_Number = TypeVar("_Number", bound=number)
+
+# In principle the SupportsInt, SupportsFloat, etc. protocols
+# would be more suited here but they have one significant disadvatange:
+# all these methods are also defined available in ndarray,
+# which has no guarantee of being scalar-esque (i.e. a 0D array)
+_ScalarNumeric = Union[int, float, complex, bool, bool_, number]
```

Comment:
Not the most ideal solution (due to the lack of duck-typing) but it serves is purpose.